src/py/flwr/cli/log.py

```diff
@@ -0,0 +1,193 @@
+    """Stream logs from the beginning of a run with connection refresh."""
+
+
+# pylint: disable=unused-argument
+def print_logs(run_id: int, channel: grpc.Channel, timeout: int) -> None:
+    """Print logs from the beginning of a run."""
+
+
+def log(
+    run_id: Annotated[
+        int,
+        typer.Argument(help="The Flower run ID to query"),
+    ],
+    directory: Annotated[
+        Path,
+        typer.Argument(help="Path of the Flower project to run"),
+    ] = Path("."),
+    federation_name: Annotated[
+        Optional[str],
+        typer.Argument(help="Name of the federation to run the app on"),
+    ] = None,
+    stream: Annotated[
+        bool,
+        typer.Option(
+            "--stream/--show",
+            help="Flag to stream or print logs from the Flower run",
+        ),
+    ] = True,
+) -> None:
+    """Get logs from a Flower project run."""
+    typer.secho("Loading project configuration... ", fg=typer.colors.BLUE)
+
+    pyproject_path = directory / "pyproject.toml" if directory else None
+    config, errors, warnings = load_and_validate(path=pyproject_path)
+
+    if config is None:
+        typer.secho(
+            "Project configuration could not be loaded.\n"
+            "pyproject.toml is invalid:\n"
+            + "\n".join([f"- {line}" for line in errors]),
+            fg=typer.colors.RED,
+            bold=True,
+        )
+        sys.exit()
+
+    if warnings:
+        typer.secho(
+            "Project configuration is missing the following "
+            "recommended properties:\n" + "\n".join([f"- {line}" for line in warnings]),
+            fg=typer.colors.RED,
+            bold=True,
+        )
+
+    typer.secho("Success", fg=typer.colors.GREEN)
+
+    federation_name = federation_name or config["tool"]["flwr"]["federations"].get(
+        "default"
+    )
+
+    if federation_name is None:
+        typer.secho(
+            "❌ No federation name was provided and the project's `pyproject.toml` "
+            "doesn't declare a default federation (with a SuperExec address or an "
+            "`options.num-supernodes` value).",
+            fg=typer.colors.RED,
+            bold=True,
+        )
+        raise typer.Exit(code=1)
+
+    # Validate the federation exists in the configuration
+    federation = config["tool"]["flwr"]["federations"].get(federation_name)
+    if federation is None:
+        available_feds = {
+            fed for fed in config["tool"]["flwr"]["federations"] if fed != "default"
+        }
+        typer.secho(
+            f"❌ There is no `{federation_name}` federation declared in the "
+            "`pyproject.toml`.\n The following federations were found:\n\n"
+            + "\n".join(available_feds),
+            fg=typer.colors.RED,
+            bold=True,
+        )
+        raise typer.Exit(code=1)
+
+    if "address" in federation:
+        _log_with_superexec(federation, run_id, stream)
+    else:
+        pass
```

Comment:
Should we log something in this case? like a message saying that `flwr logs` only makes sense when the run runs via a superexec? I presume this line would only be hit a user that passes a made up `run_id`?

Reply:
For example if i go to `examples/quickstart-pytorch` and do:
```bash
flwr log 1234
# the below prints
Loading project configuration... 
Success
```

Reply:
Actually, when all the PRs are merged, any made up `run_id` will throw an error. But it does happen later when a request is sent to the `SuperExec`. I've added a `sys.exit()` and a `typer` error for now.

Reply:
but what about if `address` is not defined in the `[federation]` block of the `pyproject.toml`? there won't be `SuperExec` to interface with in that case

Reply:
Yes, the `sys.exit()` happens in the `else` block. Specifically, I've update it to:
```python
    if "address" in federation:
        _log_with_superexec(federation, run_id, stream)
    else:
        typer.secho(
            "❌ `flwr log` currently works with `SuperExec`. Ensure that the correct"
            "`SuperExec` address is provided in the `pyproject.toml`.",
            fg=typer.colors.RED,
            bold=True,
        )
        sys.exit()
```